server/src/main/java/org/opensearch/action/admin/cluster/shards/routing/weighted/get/TransportGetWeightedRoutingAction.java

```diff
@@ -0,0 +1,114 @@
+    private final WeightedRoutingService weightedRoutingService;
+
+    @Inject
+    public TransportGetWeightedRoutingAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        WeightedRoutingService weightedRoutingService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        super(
+            ClusterGetWeightedRoutingAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            ClusterGetWeightedRoutingRequest::new,
+            indexNameExpressionResolver
+        );
+        this.weightedRoutingService = weightedRoutingService;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected ClusterGetWeightedRoutingResponse read(StreamInput in) throws IOException {
+        return new ClusterGetWeightedRoutingResponse(in);
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(ClusterGetWeightedRoutingRequest request, ClusterState state) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
+    }
+
+    @Override
+    protected void clusterManagerOperation(
+        final ClusterGetWeightedRoutingRequest request,
+        ClusterState state,
+        final ActionListener<ClusterGetWeightedRoutingResponse> listener
+    ) throws IOException {
+        try {
+            weightedRoutingService.verifyAwarenessAttribute(request.getAwarenessAttribute());
+            WeightedRoutingMetadata weightedRoutingMetadata = state.metadata().custom(WeightedRoutingMetadata.TYPE);
+            ClusterGetWeightedRoutingResponse clusterGetWeightedRoutingResponse = new ClusterGetWeightedRoutingResponse();
+            String weight = null;
+            if (weightedRoutingMetadata != null && weightedRoutingMetadata.getWeightedRouting() != null) {
+                WeightedRouting weightedRouting = weightedRoutingMetadata.getWeightedRouting();
+                if (request.local()) {
+                    DiscoveryNode localNode = state.getNodes().getLocalNode();
+                    if (localNode.getAttributes().get(request.getAwarenessAttribute()) != null) {
+                        String attrVal = localNode.getAttributes().get(request.getAwarenessAttribute());
+                        if (weightedRouting.weights().containsKey(attrVal)) {
+                            weight = weightedRouting.weights().get(attrVal).toString();
+                        }
+                    }
+                }
+                clusterGetWeightedRoutingResponse = new ClusterGetWeightedRoutingResponse(weight, weightedRouting);
+            }
+            listener.onResponse(clusterGetWeightedRoutingResponse);
+        } catch (ActionRequestValidationException ex) {
```

Comment:
we are throwing IOException what happens if we run into IOException?

Reply:
returns 500 response in case of IOException

Reply:
I meant will the listener be invoked or the request will be deemed stuck

Reply:
Yeah correct, request can get stuck on this. Made change to invoke listener on all failures